### ISSUE TITLE
Support constants where we support literals

### ISSUE THREAD

**Issue by dominikh** [dominikh/go-tools#513]
We have various checks that point out incorrect or useless code involving literals. We only support literals, not constants, because constants may differ depending on build tags.

As a trivial example, flagging `x * 0` is simple, it always has the same result and isn't very useful. Flagging `x * y`, however, will be prone to false positives if the value of y differs based on build tags. It might be `2` under one configuration and `0` under another. Suddenly, the presence of multiplication by 0 is useful: it conditionally eliminates the x term.

We can't easily get all possible values for a constant. We'd have to type-check the code for every possible combination of build tags. We can, however, deduce whether a constant's definition could possibly be affected by build tags. If the file a constant is defined in doesn't have `+build` tags, nor GOOS/GOARCH elements in the file name, nor uses cgo, then the file is unconditionally part of every build, and no other definition of the constant can exist.

**Comment by horacimacias** [comment]
can you please point me to some existing code where you're checking for literals being used?

In my case I'd like to detect strings.ReplaceAll being used with the wrong order of arguments, e.g. 
strings.ReplaceAll("oldtext", "newtext", mystring)

this is most likely a wrong usage and should be instead 
strings.ReplaceAll(mystring, "oldtext", "newtext")

(replacing on a constant/literal string seems quite useless unless I'm missing something)

Long story short, would you have any suggestions of code to look at to try and implement this check?

**Comment by dominikh** [comment]
In general I would recommend reading https://github.com/golang/example/tree/master/gotypes
